### PR TITLE
Integrate two-level ADMM with `AdmmEnv`

### PR DESCRIPTION
**Pull request by kibaekkim**
Tested on cpu and gpu with case9

### CODE REVIEW

File: src/admm/environment.jl [hunk 1]
```diff
@@ -164,16 +236,18 @@ mutable struct AdmmEnv{T,TD,TI,TM}
...
-
+        env.model = Model{T,TD,TI}(env.data, use_gpu, use_polar, use_twolevel)
         ybus = Ybus{Array{T}}(computeAdmitances(
             env.data.lines, env.data.buses, env.data.baseMVA; VI=Array{Int}, VD=Array{T})...)
 
-        init_solution!(env.solution, env.data, ybus, env.model.gen_start, env.model.line_start,
-                    rho_pq, rho_va, wRIij)
+        if !use_twolevel
+            env.solution = Solution{T,TD}(env.model)
+            init_solution!(env, ybus, rho_pq, rho_va)
```
Comment (frapac) [inline src/admm/environment.jl:245]: maybe we could use dispatch over `Solution` here? Something like
```julia
init_solution!(env, solution, ybus, rho_pq, rho_va)
```

**Review by frapac — CHANGES_REQUESTED** [review]
It overall looks good to me! I think this will ease the integration of ProxAL inside ExaTron. 
Once this PR merged, it would be nice to test properly the two level algorithm. Once https://github.com/exanauts/ExaTron.jl/pull/24/ is merged, I could take care of that.

File: src/admm/acopf_admm_gpu.jl [hunk 2]
```diff
@@ -162,8 +162,8 @@ function init_solution!(sol, data, ybus, gen_start, line_start,
         u_curr[pij_idx+6] = 0.0
         u_curr[pij_idx+7] = 0.0
         =#
-        wRIij[2*(l-1)+1] = wR0
-        wRIij[2*l] = 0.0
+        # wRIij[2*(l-1)+1] = wR0
```
[the remark this answers is not in the document]
Comment (youngdae) [reply src/admm/acopf_admm_gpu.jl:165]: wRIij is used only for auglag_kernel(). Unless we plan to support auglag_kernel() in our current or future code, we could get rid of this variable and delete the call to auglag_kernel() in our code as well.

File: src/admm/environment.jl [hunk 3]
```diff
@@ -19,6 +19,13 @@ mutable struct Parameters
     eta::Float64        # TODO: not used
     verbose::Int
 
+    # Two-Level ADMM
+    outer_eps::Float64
+    Kf::Int             # TODO: not used
```
[the remark this answers is not in the document]
Comment (youngdae) [reply src/admm/environment.jl:24]: `Kf` was for adaptive ADMM parameter tuning. We do not activate it for now.

**Review by youngdae — APPROVED** [review]
I agree with Francois' comments. Other than that, it looks good to me.

File: src/admm/environment.jl (hunk 1 above)
Comment (kibaekkim) [reply src/admm/environment.jl]: done

File: src/admm/environment.jl [hunk 4]
```diff
@@ -132,6 +161,48 @@ mutable struct Solution{T,TD}
     end
 end
 
+"""
+    Solution2{T,TD}
+
+This contains the solutions of ACOPF model instance for two-level ADMM algorithm,
+    including the ADMM parameter rho.
+"""
+mutable struct Solution2{T,TD} <: AbstractSolution{T,TD}
```
[the remark this answers is not in the document]
Comment (kibaekkim) [reply src/admm/environment.jl]: done

File: src/admm/environment.jl [hunk 5]
```diff
@@ -95,16 +113,27 @@ mutable struct Model{T,TD,TI}
...
 
+abstract type AbstractSolution{T,TD} end
+
 """
     Solution{T,TD}
 
 This contains the solutions of ACOPF model instance, including the ADMM parameter rho.
 """
-mutable struct Solution{T,TD}
+mutable struct Solution{T,TD} <: AbstractSolution{T,TD}
```
[the remark this answers is not in the document]
Comment (kibaekkim) [reply src/admm/environment.jl]: done

File: src/admm/acopf_admm_gpu.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (kibaekkim) [reply src/admm/acopf_admm_gpu.jl:163]: I will just leave them for now. @youngdae you can decide whether to remove or not later.

**Review by frapac — APPROVED** [review]
Looks good to me!

**Pull request merged**